File: app/src/main/kotlin/net/tigr/soulcalc/domain/parser/LineClassifier.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Albert Zenkoff
 */

package net.tigr.soulcalc.domain.parser

/**
 * Classifies input lines before parsing.
 *
 * This allows early detection of comments and empty lines,
 * avoiding unnecessary parsing overhead.
 */
object LineClassifier {

    /**
     * Types of lines that can be identified.
     */
    enum class LineType {
        /** Line starts with # - a comment, not evaluated */
        COMMENT,

        /** Line is empty or contains only whitespace */
        EMPTY,

        /** Line contains an expression to be parsed and evaluated */
        EXPRESSION
    }

    /**
     * Classifies the given input line.
     *
     * @param input The raw input string
     * @return The type of line
     */
    fun classify(input: String): LineType {
        val trimmed = input.trim()

        return when {
            trimmed.isEmpty() -> LineType.EMPTY
            trimmed.startsWith("#") -> LineType.COMMENT
            else -> LineType.EXPRESSION
        }
    }

    /**
     * Returns true if the line should be parsed and evaluated.
     */
    fun shouldEvaluate(input: String): Boolean {
        return classify(input) == LineType.EXPRESSION
    }

    /**
     * Returns true if the line is a comment.
     */
    fun isComment(input: String): Boolean {
        return classify(input) == LineType.COMMENT
    }

    /**
     * Returns true if the line is empty or whitespace-only.
     */
    fun isEmpty(input: String): Boolean {
        return classify(input) == LineType.EMPTY
    }
}

File: ios/SoulCalc/Domain/Parser/LineClassifier.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Albert Zenkoff

import Foundation

/// Types of lines that can be identified.
enum LineType {
    /// Line starts with # - a comment, not evaluated
    case comment

    /// Line is empty or contains only whitespace
    case empty

    /// Line contains an expression to be parsed and evaluated
    case expression
}

/// Classifies input lines before parsing.
enum LineClassifier {

    /// Classifies the given input line.
    static func classify(_ input: String) -> LineType {
        let trimmed = input.trimmingCharacters(in: .whitespaces)

        if trimmed.isEmpty {
            return .empty
        } else if trimmed.hasPrefix("#") {
            return .comment
        } else {
            return .expression
        }
    }

    /// Returns true if the line should be parsed and evaluated.
    static func shouldEvaluate(_ input: String) -> Bool {
        classify(input) == .expression
    }

    /// Returns true if the line is a comment.
    static func isComment(_ input: String) -> Bool {
        classify(input) == .comment
    }

    /// Returns true if the line is empty or whitespace-only.
    static func isEmpty(_ input: String) -> Bool {
        classify(input) == .empty
    }
}
